### ISSUE TITLE
Fixing: Unacceptable TLS error

### ISSUE THREAD

**Issue by jonathon-love** [flathub/flathub#2580]
hi,

this is a dupe of #502, and i think relevant to #2123

the command `flatpak remote-add --user --if-not-exists flathub https://flathub.org/repo/flathub.flatpakrepo` fails on appveyor, with `error: Can't load uri https://flathub.org/repo/flathub.flatpakrepo: Unacceptable TLS certificate` i'm using the `Ubuntu2004` image.

there is some discussion of how this might be fixed on #502, however there seems to be assumed knowledge as to what is actually failing here, which i don't possess. any tips on how this might be addressed?

i note that it's possible to install the flathub remote as a by-product of perfoming this sort of install:

```
flatpak install --user -y https://dl.flathub.org/build-repo/....;
```

i can install the remote by performing an installation as the above, then uninstall that, then i can install from flathub with:

```
flatpak install flathub ...
```

with thanks

**Comment by jonathon-love** [comment]
hi,

i think i've figured this out. this stems from the let's encrypt expiry. if i copy `flathub.flatpakrepo` to a server not protected with a let's encrypt certificate, the following succeeds:

```
flatpak remote-add --user --if-not-exists flathub https://my-server-here/path/to/flathub.flatpakrepo
```

if flathub can update to a non-let's encrypt cert, this will fix things for people using older OSes, etc.

**Comment by mkrajnak** [comment]
I can reproduce this on older RHEL installation (flatpak-1.0.9-12) and I found a workaround which is installing flathub through gnome-software.

**Comment by mcatanzaro** [comment]
Can you try to figure out which of curl's TLS backends flatpak is using on Ubuntu? As far as I can tell, I *think* the default is OpenSSL, which is the same we have on Fedora. But there should be no way OpenSSL should be failing with this simple chain unless it is way older than you have in Ubuntu 20.04.

Please paste the output of:

 * `curl -I https://flathub.org`
 *  `openssl s_client -verify_return_error -connect flathub.org:443`
 *  For fun: `gnutls-cli flathub.org`

Your distro's command-not-found handler should be able to install these tools for you if you are missing them.

> I can reproduce this on older RHEL installation (flatpak-1.0.9-12) and I found a workaround which is installing flathub through gnome-software.

Not good. What version of RHEL?

**Comment by mcatanzaro** [comment]
> Can you try to figure out which of curl's TLS backends flatpak is using on Ubuntu? 

I mixed this up with #2123. Suffice to say: distro is very relevant. The curl TLS backend used varies by distro and by distro version. RHEL 8 should be openssl. I think RHEL 7 would be... NSS? Or perhaps GnuTLS? I'm really not sure. Would have to check to be certain.

**Comment by mcatanzaro** [comment]
Also check to ensure you have ISRG Root X1 in your trust store. How to do this varies by distro. Using seahorse would be a safe bet that should work everywhere.

**Comment by kerivkennedy** [comment]
Hopefully someone can help me with this.  I'm a total newb to linux  
trying to install flatpak .  I want to install Blender on my new Chromebook.  I installed the flatpak and attempted the next step (flathub repository) and get the same error TLS certificate.  

hopefully someone can explain in super easy terms (this was a Christmas present, so I;m just learning this. back in the mid 90s I thought nothing of using DOS commands to do everything on a computer.  then windows went and made everything to easy.  So I am not computer illiterate, just have been dumbed-down . 

I know my chromebook is Lenovo and is using the ARM  (my husband has the pixel slate and has helped trying to set up Prusa Slicer)

**Comment by hfiguiere** [comment]
Blender is not available ARM so it's gonna be pointless.

**Comment by jonathon-love** [comment]
> Also check to ensure you have ISRG Root X1 in your trust store.

so i think there's two audiences interested in this issue ... 1. is people trying to install software, 2. is people publishing software.

for 1. i'm like "sure, i don't mind noodling around in my computer, figuring out what this seahorse thing is", etc.

for 2. i'm like ... "wait? so exactly what do i communicate to my [often complete novice] users?"

the exciting thing about flatpaks is that we can make software available to chromebooks, a whole class of user traditionally cut off from the linux ecosystem.

so it would be great if these sorts of issues could generate a more sympathetic response, and perhaps even a solution. it's easy to dismiss novice users, but i think there's real value in making things accessible to as broad an audience as possible.

a reasonably simple solution here would be to use a non-let's encrypt cert, or if that's too much, simply mirroring the `flathub.flatpakrepo` file on a separate server using a different cert, and instructing people to install from there instead.

with thanks

**Comment by mcatanzaro** [comment]
> so it would be great if these sorts of issues could generate a more sympathetic response, and perhaps even a solution. it's easy to dismiss novice users, but i think there's real value in making things accessible to as broad an audience as possible.

I replied above with a request for several bits of info, which you haven't provided yet. Maybe you missed that? This issue is blocked on you, not on devs. If you provide the requested info, I'll take a look.

> a reasonably simple solution here would be to use a non-let's encrypt cert,

No, we need to figure out what is actually wrong.

**Comment by mcatanzaro** [comment]
> Also check to ensure you have ISRG Root X1 in your trust store. How to do this varies by distro. Using seahorse would be a safe bet that should work everywhere.

seahorse is a graphical application (possibly still installed by default on Ubuntu, last I checked?) so it should be pretty simple, but this step can be optional, as it's unlikely that you would be missing this cert.

**Comment by jonathon-love** [comment]
this is from the focal image running on appveyor

```
curl -I https://flathub.org
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0  7451    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0HTTP/2 200 
server: nginx/1.18.0 (Ubuntu)
date: Mon, 27 Dec 2021 01:42:23 GMT
content-type: text/html
content-length: 7451
vary: Accept-Encoding
last-modified: Tue, 28 Sep 2021 14:41:06 GMT
vary: Accept-Encoding
etag: "61532982-1d1b"
accept-ranges: bytes
expires: Wed, 26 Jan 2022 01:42:23 GMT
cache-control: max-age=2592000
cache-control: public, 
o-transform
```

```
openssl s_client -verify_return_error -connect flathub.org:443
depth=2 C = US, O = Internet Security Research Group, CN = ISRG Root X1
verify return:1
depth=1 C = US, O = Let's Encrypt, CN = R3
verify return:1
depth=0 CN = flathub.org
verify return:1
DONE
CONNECTED(00000003)
---
Certificate chain
 0 s:CN = flathub.org
   i:C = US, O = Let's Encrypt, CN = R3
 1 s:C = US, O = Let's Encrypt, CN = R3
   i:C = US, O = Internet Security Research Group, CN = ISRG Root X1
 2 s:C = US, O = Internet Security Research Group, CN = ISRG Root X1
   i:O = Digital Signature Trust Co., CN = DST Root CA X3
---
Server certificate
-----BEGIN CERTIFICATE-----
MIIGSjCCBTKgAwIBAgISA2hVQJ+XublatKmkBpXh7vpVMA0GCSqGSIb3DQEBCwUA
MDIxCzAJBgNVBAYTAlVTMRYwFAYDVQQKEw1MZXQncyBFbmNyeXB0MQswCQYDVQQD
EwJSMzAeFw0yMTExMjcxNjQ4MDdaFw0yMjAyMjUxNjQ4MDZaMBYxFDASBgNVBAMT
C2ZsYXRodWIub3JnMIICIjANBgkqhkiG9w0BAQEFAAOCAg8AMIICCgKCAgEAt+7n
wacOzjEIMVNuqHvkqTdA/8Dti16ov6neWr7kLi4dO5bp4TQzszeMhJAg7USuZyPT
tTIIpNb6kOyCBVXTcPSOWQ1FYeHyOQW37GMdZMgwbB+/WTvzWOopyo2ZnMH+UOPt
raLLAkOD1Z7+2Qp/LTrgjzW5KbfpDCvoB5fO3UtMRypYpgKE0cLye9A68VsVCMe9
tvSIT0x3h5b8q4m4ES5dKpj7Djc6a08DOlCk52C0lh3Pj9EEhpbhqU3HLqn8Napg
MC2akq2qaf2WYeL1CpaAcAT4F7Owhw7B+mCpBhIYGt5Wdn0wL8M+3tQeHrLwE1nw
vgbtnMHEvH5wuPTqmMBwpdad2VjZ4wMT2WlfTVi7UMWkPzZjXObc0TViUaXky9yE
sGA7ZIHrb+zzCIqraV5+NgboJ6lU3JzyirUTOb0jRPEefj4Jg8SX9SruOoW/gGBC
ArHqFADECLUM7eVUwKE/SF4SleO97JB3t3+7/3oN2RFbeUQQwiZlioAkmqvp9kQ9
hChT5OS9tqKtvM1Mq04kBbB0eCUzIOAgMRh90xYb5pWV4Kl30wJ9QIlDOt5oiUqr
3BAefv1pNgqGirhhsDXSOTUjaYH+7gp68duWw6xj0fKjhGB1jwWUgs1/dobM9yIV
m9SyoZqVWSTcUsW76tm+pbi96CboCw969cXnJ9MCAwEAAaOCAnQwggJwMA4GA1Ud
DwEB/wQEAwIFoDAdBgNVHSUEFjAUBggrBgEFBQcDAQYIKwYBBQUHAwIwDAYDVR0T
AQH/BAIwADAdBgNVHQ4EFgQUGaHdM4ymQW3cTiLALicxIhzUwsEwHwYDVR0jBBgw
FoAUFC6zF7dYVsuuUAlA5h+vnYsUwsYwVQYIKwYBBQUHAQEESTBHMCEGCCsGAQUF
BzABhhVodHRwOi8vcjMuby5sZW5jci5vcmcwIgYIKwYBBQUHMAKGFmh0dHA6Ly9y
My5pLmxlbmNyLm9yZy8wRAYDVR0RBD0wO4ILZmxhdGh1Yi5vcmeCG2ZsYXRodWIu
d2ViYXBwcy5mbGF0aHViLm9yZ4IPd3d3LmZsYXRodWIub3JnMEwGA1UdIARFMEMw
CAYGZ4EMAQIBMDcGCysGAQQBgt8TAQEBMCgwJgYIKwYBBQUHAgEWGmh0dHA6Ly9j
cHMubGV0c2VuY3J5cHQub3JnMIIBBAYKKwYBBAHWeQIEAgSB9QSB8gDwAHcA36Ve
q2iCTx9sre64X04+WurNohKkal6OOxLAIERcKnMAAAF9YoLsFwAABAMASDBGAiEA
mah6pmanMfd1QRm9MEZ8UmBhVDom6xaRfwTPj7pOn9QCIQCIHYW55/ruTYG7gN4n
nuHVoEaXxdVRwrMHVv9WlrGF/AB1AEalVet1+pEgMLWiiWn0830RLEF0vv1JuIWr
8vxw/m1HAAABfWKC7EMAAAQDAEYwRAIgKYYHJmws7alhezlTvQW6Qd3amZXc5vS8
GSvyvsXjL/YCIGp67kw8fwH0NAI8sW7sUBkd3vRiv4RYXNsDgR2ShkgtMA0GCSqG
SIb3DQEBCwUAA4IBAQCchy/xw5k3sQIvgJAO7I6EMB+RYkB/Frtod8CYj4rd5UVT
MAflOuel/hWQ0n9P2vcmSOvQrclowwi/FlnSaIpYfsHNMAo/+77ox7aNVRa9lfL/
OYoEM8jrtJAxgkRtlc4FCDA9XNGB80wQ/l+0+tu9SvYsuJzoRRJ/oxfdWf0NYM63
11yTBmRsh/CeJ9YF7pcqGcnJQB5Dlta+uDVNN4d7bLJVTki95qen0E1+yY4jRyM7
lc2fay3E3/Q3W9zNdy7uUvUgG7JCb7F0+zAEnrD8q/gIn+3Z5wwNCLJ9ewlBX374
RjQAqNt4sOjJkAAvU2JKv1t55xwB/kv6NTVsPyxl
-----END CERTIFICATE-----
subject=CN = flathub.org
issuer=C = US, O = Let's Encrypt, CN = R3
---
No client certificate CA names sent
Peer signing digest: SHA256
Peer signature type: RSA-PSS
Server Temp Key: X25519, 253 bits
---
SSL handshake has read 5126 bytes and written 383 bytes
Verification: OK
---
New, TLSv1.3, Cipher is TLS_AES_256_GCM_SHA384
Server public key is 4096 bit
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)
---
```
```
gnutls-cli flathub.org
Processed 128 CA certificate(s).
Resolving 'flathub.org:443'...
Connecting to '46.235.230.111:443'...
- Certificate type: X.509
- Got a certificate list of 3 certificates.
- Certificate[0] info:
 - subject `CN=flathub.org', issuer `CN=R3,O=Let's Encrypt,C=US', serial 0x036855409f97b9b95ab4a9a40695e1eefa55, RSA key 4096 bits, signed using RSA-SHA256, activated `2021-11-27 16:48:07 UTC', expires `2022-02-25 16:48:06 UTC', pin-sha256="8UTCF0B/aVq34SPe6dHfibtHlmhJqeCpKre23k1yckc="
	Public Key ID:
		sha1:9170621f8b1c2b5301824069beffebb1542da8ce
		sha256:f144c217407f695ab7e123dee9d1df89bb47966849a9e0a92ab7b6de4d727247
	Public Key PIN:
		pin-sha256:8UTCF0B/aVq34SPe6dHfibtHlmhJqeCpKre23k1yckc=
- Certificate[1] info:
 - subject `CN=R3,O=Let's Encrypt,C=US', issuer `CN=ISRG Root X1,O=Internet Security Research Group,C=US', serial 0x00912b084acf0c18a753f6d62e25a75f5a, RSA key 2048 bits, signed using RSA-SHA256, activated `2020-09-04 00:00:00 UTC', expires `2025-09-15 16:00:00 UTC', pin-sha256="jQJTbIh0grw0/1TkHSumWb+Fs0Ggogr621gT3PvPKG0="
- Certificate[2] info:
 - subject `CN=ISRG Root X1,O=Internet Security Research Group,C=US', issuer `CN=DST Root CA X3,O=Digital Signature Trust Co.', serial 0x4001772137d4e942b8ee76aa3c640ab7, RSA key 4096 bits, signed using RSA-SHA256, activated `2021-01-20 19:14:03 UTC', expires `2024-09-30 18:14:03 UTC', pin-sha256="C5+lpZ7tcVwmwQIMcRtPbsQtWLABXhQzejna0wHFr8M="
- Status: The certificate is trusted. 
- Description: (TLS1.3-X.509)-(ECDHE-SECP256R1)-(RSA-PSS-RSAE-SHA256)-(AES-256-GCM)
- Options: OCSP status request,
- Handshake was completed
- Simple Client Mode:
- Peer has closed the GnuTLS connection
```

**Comment by mcatanzaro** [comment]
Er, but all of those succeeded. Hmm. Please try your original failing command again to confirm it is really still failing:

```
$ flatpak remote-add --user --if-not-exists flathub https://flathub.org/repo/flathub.flatpakrepo
```

Is it really still broken? If only flatpak fails to handle flathub.org, that would be weird, but would certainly indicate a flatpak bug.

I just noticed this is the flathub issue tracker. This issue should be closed as there is obviously nothing wrong with flathub here. We can continue here to figure out which component is really to blame, then move to a more appropriate location.

**Comment by mcatanzaro** [comment]
> I just noticed this is the flathub issue tracker. This issue should be closed as there is obviously nothing wrong with flathub here.

BTW I should clarify: having examined the TLS certificate chain, it's clear that flathub.org is presenting a good chain. Whatever is going wrong is a bug on the client side software (ca-certificates -> openssl -> libcurl -> flatpak), not on flathub's end.

That's not to say that this was a bad place to _originally_ report the issue, but that we now know enough to confidently say it's not the right place to fix it.

**Comment by Aditya-ds-1806** [comment]
Trying to install a `flatpak` on RHEL 7, facing the same problem. Was there a fix found?

**Comment by Aditya-ds-1806** [comment]
@mkrajnak Can you elaborate what you did? I am on RHEL 7 and flatpak 1.0.2

**Comment by mcatanzaro** [comment]
> Trying to install a flatpak on RHEL 7, facing the same problem. Was there a fix found?

No, the last action in this issue was a request for me for further info. Unless you have the *exact* same problem (have run the same commands requested above and verified your output is the same) and can provide the last requested info, please don't assume this issue is really the same as yours. I doubt it, because this ticket is about Ubuntu 20.04, which is way different from RHEL 7. libcurl on RHEL 7 uses NSS, so it's a completely different environment from anywhere else.

P.S. There is no way flatpak applications will work properly on RHEL 7 because the host system is just not new enough. You're going to be missing p11-kit-server, xdg-desktop-portal[-gtk], and more. So honestly, I don't recommend even trying to get it to work. But if you really want to try, I would create a separate issue report for starters.

**Comment by Aditya-ds-1806** [comment]
@mcatanzaro I was able to install flatpak on a rhel 7 using `flatpak remote-add --user --if-not-exists flathub https://dl.flathub.org/repo/flathub.flatpakrepo`.  I also feel it is a client-side issue as I had to clear the entries in `/home/<user>/.local/share/flatpak/repo/config` after which it seemed to have worked.

**Comment by mcatanzaro** [comment]
> I also feel it is a client-side issue as I had to clear the entries in /home/<user>/.local/share/flatpak/repo/config, after which it seemed to have worked.

Er... are you sure you simply didn't have it configured already?

If you deleted the configuration that was causing the command to fail without saving it somewhere, then of course you've probably lost the reproducer required for a bug report....

**Comment by Aditya-ds-1806** [comment]
It was on a virtualbox VM. I can try to reproduce the problem on a new VM and get back to you. @mcatanzaro

**Comment by nanonyme** [comment]
My hunch is the original case has something to do with https://letsencrypt.org/docs/dst-root-ca-x3-expiration-september-2021/ but I can't figure out how.

**Comment by mcatanzaro** [comment]
> My hunch is the original case has something to do with https://letsencrypt.org/docs/dst-root-ca-x3-expiration-september-2021/ (and the fact that the second root CA expired) but I can't figure out how.

That was my first thought too, but it shouldn't be happening because Ubuntu 20.08 has the ISRG Root X1 certificate (we know this because all the commands that I asked for succeeded). And the OpenSSL version in Ubuntu 20.08 is also new enough to handle the chain of trust.

IMO there's nothing more to do here besides to wait for Jonathon to confirm whether the original command is really still failing. Maybe there really *was* a bug with flathub.org, but it was transient and already long since fixed? Who knows.

> It was on a virtualbox RHEL 7.6 VM. I can try to reproduce the problem on a new VM and get back to you.

If you're a Red Hat customer, you can report this downstream using a customer case, and the NSS developers might be willing to investigate. I see the ca-certificates package in RHEL 7.6 has been upgraded recently, so it ought to have ISRG Root X1, but maybe the old version of NSS is tripping over the expiration of DST Root CA X3. I certainly wouldn't *expect* this to work due to the very unusual configuration and old software versions, so from an upstream perspective, I wouldn't worry about this any further unless you can reproduce in RHEL 8, where things are much more normal (curl using a recent version of OpenSSL) and would be expected to work. If you really really need things to work in RHEL 7, I would try rebuilding the libcurl package to use OpenSSL instead of NSS, which ought to be able to handle this chain fine. One thing's for sure: it's *extremely* unlikely to be the same problem that Jonathon is experiencing, since you are using entirely separate distros with different curl backends and different ca-certificate packages.

**Comment by jonathon-love** [comment]
> IMO there's nothing more to do here besides to wait for Jonathon to confirm whether the original command is really still failing.

hi, apologies for the delay,

yes, the command seems to be now working. i'll touch base with some of my users and see if it is resolved for them now too.

i'd close this issue, but i'm not quite sure if i should with the subsequent discussion ... so i'll leave it to you guys.

thanks for you help.

**Comment by nanonyme** [comment]
Something I noticed: libsoup in Debian does not pull in ca-certificates while libcurl does. Curl package also does. So most likely in various installations by the time you start debugging CA issues, they become fixed through testing.

**Comment by mcatanzaro** [comment]
> i'd close this issue, but i'm not quite sure if i should with the subsequent discussion ... so i'll leave it to you guys.

It should be closed.

**Comment by nanonyme** [comment]
@jonathon-love yes, please, close this. The ca-certificates dependency mess is something Debian should resolve.

**Comment by iamihgam** [comment]
This is what I did and It resolved the problem for me:
$ sudo dpkg-reconfigure ca-certificates
On entering please deselect "DST Root CA X3". save and run the flathub repo script. It works . 
Please see: 
https://serverfault.com/questions/1079199/client-on-debian-9-erroneously-reports-expired-certificate-for-letsencrypt-issue

**Comment by mcatanzaro** [comment]
> On entering please deselect "DST Root CA X3". save and run the flathub repo script. It works .

Interesting! This should definitely not be required. That certificate should be ignored because it is expired. I think you have proved that there is a Debian bug here. If you are using a supported version of Debian (either Buster or Bullseye) then you should (1) determine whether there is a problem with libcurl's OpenSSL backend, or a problem with OpenSSL itself, using the commands I've provided [here](https://github.com/flathub/flathub/issues/2580#issuecomment-956448280), then (2) report a bug against the appropriate component on Debian's bug tracker.

My suspicion is that somehow Debian's OpenSSL is too old, and there is nothing wrong with the ca-certificates. But it's hard to be certain, since there are now three different users in this issue with at least two different problems, which is confusing things. Moving to downstream bugtracker is certainly required now.

**Comment by nanonyme** [comment]
FWIW Debian 9 -> https://packages.debian.org/stretch/libcurl3 OpenSSL 1.0.2d.

**Comment by mcatanzaro** [comment]
> OpenSSL 1.0.2d.

OK, wow, you are right. Debian 9 (Stretch) is expected to be broken, then. I got confused because I saw the default version of OpenSSL in Stretch was 1.1.0l, but that doesn't matter because curl is using the older version of OpenSSL.

You need at least OpenSSL 1.1.0 to use Flathub.

**Comment by nanonyme** [comment]
Well, I would even state you need Debian 10 or higher to use the Internet with anything curl-based given how common Let's Encrypt is.

**Comment by SamsonovAnton** [comment]
In case someone runs an elderly version of openSUSE, the solution is:

```
sudo rm /usr/share/pki/trust/DST_Root_CA_X3.pem
sudo update-ca-certificates
```

Be sure *not* to replace the existing ISRG_Root_X1.pem with a freshly downloaded version.

**Comment by nanonyme** [comment]
@SamsonovAnton is that yet another bug in crypto library that fails to ignore expired CA?

**Comment by SamsonovAnton** [comment]
> is that yet another bug in crypto library that fails to ignore expired CA?

This is the same bug as referred to by @iamihgam in the [comment above](https://github.com/flathub/flathub/issues/2580#issuecomment-1004991321_). Funny enough, out of all tests

    curl -I https://flathub.org
    openssl s_client -verify_return_error -connect flathub.org:443
    gnutls-cli flathub.org    # For fun.

the only one that was failing for me was the last one with `gnutls-cli`.

**Comment by bkauler** [comment]
I will throw in our experience with EasyOS. We have a fix, but it is very strange...

EasyOS has "Flapi", Flatpak Installer, and it works great. No problem with registering with flathub.org, but then I received a report from one person who got that "unacceptable TLS certificate" when try to register with flathub. 

So far, three people have reported this failure. Which is very puzzling, as it works for the rest of us.

Then one of those three stumbled upon a fix: to set the time from the Internet.

Another of those three has tried that, and it also fixed the problem.

Yeah, but why? Federico has suggested why on page 6 where we have been discussing this on the Puppy Linux Forum:

https://forum.puppylinux.com/viewtopic.php?t=9038

Yet federico has no problem with TLS anywhere else, only with flathub.

**Comment by bkauler** [comment]
Also, as this person reported, updating the ca-certificates did not fix the problem, but syncing time from Internet did:

https://itsfoss.com/unacceptable-tls-certificate-error-linux/

So, it seems that this issue has been closed prematurely. The flathub developers need to determine why this problem is occurring only with flathub.org, and why the sync time with Internet fixes it.

**Comment by hfiguiere** [comment]
>  and why the sync time with Internet fixes it.

If that's the fix then that's a problem in the client machines whose clock isn't on tim. Why it (seems to) happen only for flathub is left as an exercise to the reader. But it could just mean the clock drift cause a certificate to be invalid (future or past), and that could also happen with many other certs.

**Comment by elbaih101** [comment]
iam having a similar issue
```sh
(com.vysp3r.ProtonPlus:2): GLib-Net-WARNING **: 23:54:07.482: Failed to load TLS database: System trust contains zero trusted certificates; please investigate your GnuTLS configuration
** Message: 23:54:07.546: web.vala:18: Unacceptable TLS certificate

```

out put of the 3 command s up top

```sh
elbaih@abolebda:~$ gnutls-cli flathub.org
Processed 150 CA certificate(s).
Resolving 'flathub.org:443'...
Connecting to '199.232.57.91:443'...
- Certificate type: X.509
- Got a certificate list of 2 certificates.
- Certificate[0] info:
 - subject `CN=flathub.org', issuer `CN=GlobalSign Atlas R3 DV TLS CA 2024 Q4,O=GlobalSign nv-sa,C=BE', serial 0x01d815ac83a9077056f45831505689cd, RSA key 2048 bits, signed using RSA-SHA256, activated `2025-01-03 14:26:20 UTC', expires `2026-02-04 14:26:19 UTC', pin-sha256="Id6DLEnjLqtyRb3zoZFEFIeGNp7UfVN/UHv9/O48rls="
	Public Key ID:
		sha1:3b9ab85799533dba2c06d4e1db935d54d875dfa8
		sha256:21de832c49e32eab7245bdf3a19144148786369ed47d537f507bfdfcee3cae5b
	Public Key PIN:
		pin-sha256:Id6DLEnjLqtyRb3zoZFEFIeGNp7UfVN/UHv9/O48rls=

- Certificate[1] info:
 - subject `CN=GlobalSign Atlas R3 DV TLS CA 2024 Q4,O=GlobalSign nv-sa,C=BE', issuer `CN=GlobalSign,O=GlobalSign,OU=GlobalSign Root CA - R3', serial 0x00817cd51426f2ddf6ff588022c88554f6, RSA key 2048 bits, signed using RSA-SHA256, activated `2024-07-17 03:09:40 UTC', expires `2026-07-17 00:00:00 UTC', pin-sha256="NeNjKRsP6MwFGRS/L3VaQ+3MueSxLtCH4ZpxuQjQSdc="
- Status: The certificate is trusted. 
- Description: (TLS1.3-X.509)-(ECDHE-SECP256R1)-(RSA-PSS-RSAE-SHA256)-(AES-128-GCM)
- Session ID: D3:7E:E7:5D:B8:3D:3F:8E:2B:EC:17:40:DB:14:8A:AD:50:4A:02:D5:1F:DD:E2:AC:BD:AF:7A:03:04:D2:E2:34
- Options: OCSP status request,
- Handshake was completed

- Simple Client Mode:


```
```sh
elbaih@abolebda:~$ openssl s_client -verify_return_error -connect flathub.org:443
Connecting to 199.232.57.91
CONNECTED(00000003)
depth=2 OU=GlobalSign Root CA - R3, O=GlobalSign, CN=GlobalSign
verify return:1
depth=1 C=BE, O=GlobalSign nv-sa, CN=GlobalSign Atlas R3 DV TLS CA 2024 Q4
verify return:1
depth=0 CN=flathub.org
verify return:1
---
Certificate chain
 0 s:CN=flathub.org
   i:C=BE, O=GlobalSign nv-sa, CN=GlobalSign Atlas R3 DV TLS CA 2024 Q4
   a:PKEY: rsaEncryption, 2048 (bit); sigalg: RSA-SHA256
   v:NotBefore: Jan  3 14:26:20 2025 GMT; NotAfter: Feb  4 14:26:19 2026 GMT
 1 s:C=BE, O=GlobalSign nv-sa, CN=GlobalSign Atlas R3 DV TLS CA 2024 Q4
   i:OU=GlobalSign Root CA - R3, O=GlobalSign, CN=GlobalSign
   a:PKEY: rsaEncryption, 2048 (bit); sigalg: RSA-SHA256
   v:NotBefore: Jul 17 03:09:40 2024 GMT; NotAfter: Jul 17 00:00:00 2026 GMT
---
Server certificate
-----BEGIN CERTIFICATE-----
MIIGWTCCBUGgAwIBAgIQAdgVrIOpB3BW9FgxUFaJzTANBgkqhkiG9w0BAQsFADBY
MQswCQYDVQQGEwJCRTEZMBcGA1UEChMQR2xvYmFsU2lnbiBudi1zYTEuMCwGA1UE
AxMlR2xvYmFsU2lnbiBBdGxhcyBSMyBEViBUTFMgQ0EgMjAyNCBRNDAeFw0yNTAx
MDMxNDI2MjBaFw0yNjAyMDQxNDI2MTlaMBYxFDASBgNVBAMMC2ZsYXRodWIub3Jn
MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAymxy2OtBnj0xswWvo8Yj
oq3ktW0eZ0iY6JDVk8TWtysMm5cUKBOTdTbWq77iXwEfysLoHgvq9GtvXXq8u1Cr
hJhZnx1LCCAkaOA2HTWxtunYGE6zN3D2f10Hbm9RWv8CVm4xqtEyBSsycpBgmifV
rthTexz6G1AsOQ8PAhe6HfDbdSbz3stIYxUL8PUO35oiSIsyTMcq+JtsWynB1Rlo
xwuQa+xcQC8j1b1r25bUcjzI7fQgEOHr4EXGESCMKx/ctZpflGlx/FjWwaBNRNdB
4OghfZHk4xvwakerzuIGUPZpWHuIXiof7C6jIv2reo4O2ibXK+dMLlq1f3KiZVxQ
0wIDAQABo4IDXzCCA1swFgYDVR0RBA8wDYILZmxhdGh1Yi5vcmcwDgYDVR0PAQH/
BAQDAgWgMB0GA1UdJQQWMBQGCCsGAQUFBwMBBggrBgEFBQcDAjAdBgNVHQ4EFgQU
PIV9bP2xpLrbnXAAQyntM7ZfIZQwVwYDVR0gBFAwTjAIBgZngQwBAgEwQgYKKwYB
BAGgMgoBAzA0MDIGCCsGAQUFBwIBFiZodHRwczovL3d3dy5nbG9iYWxzaWduLmNv
bS9yZXBvc2l0b3J5LzAMBgNVHRMBAf8EAjAAMIGeBggrBgEFBQcBAQSBkTCBjjBA
BggrBgEFBQcwAYY0aHR0cDovL29jc3AuZ2xvYmFsc2lnbi5jb20vY2EvZ3NhdGxh
c3IzZHZ0bHNjYTIwMjRxNDBKBggrBgEFBQcwAoY+aHR0cDovL3NlY3VyZS5nbG9i
YWxzaWduLmNvbS9jYWNlcnQvZ3NhdGxhc3IzZHZ0bHNjYTIwMjRxNC5jcnQwHwYD
VR0jBBgwFoAUYJHsHALyDv5jT2XLYrACKgNY6bMwSAYDVR0fBEEwPzA9oDugOYY3
aHR0cDovL2NybC5nbG9iYWxzaWduLmNvbS9jYS9nc2F0bGFzcjNkdnRsc2NhMjAy
NHE0LmNybDCCAX4GCisGAQQB1nkCBAIEggFuBIIBagFoAHYADleUvPOuqT4zGyyZ
B7P3kN+bwj1xMiXdIaklrGHFTiEAAAGULI9fQAAABAMARzBFAiEA0rpaCWcX11MQ
wbraq/n2V/PEY8f77HedwrxkeEDNTMICIDb3M6egIi3ITX0WNomEqolRoFDLX114
60coMnGghQAaAHYAGYbUxyiqb/66A294Kk0BkarOLXIxD67OXXBBLSVMx9QAAAGU
LI9fjQAABAMARzBFAiEAh4kO2hkUsTXFEr/iUIiFyVNHVpd78MUVU71BBJZU2mkC
ICkJfxO6YXz5UeqoSRm2sOwOm3qFolIGc6gR2H18dfZXAHYAJS+Uwisp6W6fQRpy
BytpXFtS/5epDSVAu/zcUexN7gsAAAGULI9i5QAABAMARzBFAiBJA8JK+tp2Nl9u
4OQntN3E80UWQEJqHfiRaGEv0IhcFAIhANMDJD8/tNunHPcZEzthjgbgFdJTqBxL
0+JctBpiUfpXMA0GCSqGSIb3DQEBCwUAA4IBAQByOjSDt/nrbuw9RA88lkxJIgm4
5JiYm2fq5koiG5g3D1EDt++753oQyNdsa0Py2uT0ADwzI6OYerTGda30CfvW/NLG
QvzzjkvrOK04Gg8C3DXYi/UKVsMFYOe2pMc/2ZPuvNiR9yQ+Z8JZXJW3VFsCY+Fs
878joeyKhOm/vjq+XX5BYEG8Mh8w91Zzzh2QqNEmgOlrvK0oLI1XSygRb4N3QxBl
g4sjqBoTplPedndA0SGVV8JrY4tGC134R3GHe2AMK1uZE4gbr+7DeL1oIxj8ElzH
gbUGwqE4UsCUstP7uHAWH/UyxAbeXgVuaS0mcAM5toFry18FJnjCkKN4oobr
-----END CERTIFICATE-----
subject=CN=flathub.org
issuer=C=BE, O=GlobalSign nv-sa, CN=GlobalSign Atlas R3 DV TLS CA 2024 Q4
---
No client certificate CA names sent
Peer signing digest: SHA256
Peer signature type: RSA-PSS
Server Temp Key: X25519, 253 bits
---
SSL handshake has read 3350 bytes and written 386 bytes
Verification: OK
---
New, TLSv1.3, Cipher is TLS_AES_128_GCM_SHA256
Protocol: TLSv1.3
Server public key is 2048 bit
This TLS version forbids renegotiation.
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)
---
---
Post-Handshake New Session Ticket arrived:
SSL-Session:
    Protocol  : TLSv1.3
    Cipher    : TLS_AES_128_GCM_SHA256
    Session-ID: E8B83535799B1710FE50D5825C1C0ECD8EF3C96073C36F3983580CB89C2DE949
    Session-ID-ctx: 
    Resumption PSK: 371EC0D6C79C9FDD5156AED8DD8A5C8524A5BFB59523CA302AF51072859359EE
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    TLS session ticket lifetime hint: 86400 (seconds)
    TLS session ticket:
    0000 - 52 61 f8 9e d1 7e a5 83-15 ff 05 58 6b b9 ff 5b   Ra...~.....Xk..[
    0010 - 9c 5d 13 0f c7 84 90 7c-08 5a 47 55 f2 f4 90 b6   .].....|.ZGU....
    0020 - d9 c1 53 9b 6a 7a a0 5d-1c ae f8 00 b9 ca 11 2b   ..S.jz.].......+
    0030 - f8 6e 25 e7 60 bf b4 6b-b1 c4 f1 31 b8 87 4c 84   .n%.`..k...1..L.
    0040 - 0d c3 74 d0 52 6b 14 a5-45 1c 25 c7 b3 92 9a a5   ..t.Rk..E.%.....
    0050 - ff 59 d9 3e 8c dc 6c 58-ca fe 6f dd 6c 5a 4e fe   .Y.>..lX..o.lZN.
    0060 - 42 f6 4a e4 c6 81 03 e8-3d 10 00 47 4b a3 93 3c   B.J.....=..GK..<
    0070 - ca 7a 00 6c ed ab 2b 5f-9f a9 67 27 f6 f2 36 1e   .z.l..+_..g'..6.
    0080 - a4 62 4e 73 9c 97 2a 1d-f1 12 38 7a 0a 40 f3 77   .bNs..*...8z.@.w
    0090 - c9 68 14 a6 d5 a7 4c b6-03 5a ec 9b a1 c3 6f f3   .h....L..Z....o.

    Start Time: 1743371929
    Timeout   : 7200 (sec)
    Verify return code: 0 (ok)
    Extended master secret: no
    Max Early Data: 8192
---
read R BLOCK


```
```sh
elbaih@abolebda:~$ curl -I https://flathub.org
HTTP/2 200 
server: nginx/1.26.2
content-type: text/html; charset=utf-8
x-frame-options: Deny
x-xss-protection: 1; mode=block
x-content-type-options: nosniff
referrer-policy: strict-origin-when-cross-origin
permissions-policy: accelerometer=(), ambient-light-sensor=(), autoplay=(), battery=(), camera=(), cross-origin-isolated=(), display-capture=(), document-domain=(), encrypted-media=(), execution-while-not-rendered=(), execution-while-out-of-viewport=(), fullscreen=(), geolocation=(), gyroscope=(), keyboard-map=(), magnetometer=(), microphone=(), midi=(), navigation-override=(), payment=(self), picture-in-picture=(), publickey-credentials-get=(), screen-wake-lock=(), sync-xhr=(), usb=(), web-share=(), xr-spatial-tracking=()
content-security-policy: base-uri 'self' https://flathub.org; default-src 'none'; form-action 'none'; script-src 'self' 'sha256-yYJq8IP5/WhJj6zxyTmujEqBFs/MufRufp2QKJFU76M=' https://webstats.gnome.org https://js.stripe.com https://js.stripe.com https://maps.googleapis.com; style-src 'self' 'unsafe-inline' https://dl.flathub.org; font-src 'self'; connect-src 'self' https://flathub.org https://webstats.gnome.org https://api.stripe.com https://maps.googleapis.com https://o467221.ingest.sentry.io/api/6610580/; img-src 'self' https://dl.flathub.org https://webstats.gnome.org https://avatars.githubusercontent.com https://gitlab.com https://gitlab.gnome.org https://lh3.googleusercontent.com https://secure.gravatar.com https://invent.kde.org data:; frame-ancestors 'none'; frame-src 'self' https://*.js.stripe.com https://js.stripe.com https://hooks.stripe.com;
x-nextjs-cache: HIT
x-nextjs-prerender: 1
x-powered-by: Next.js
cache-control: s-maxage=900, stale-while-revalidate=31535100
etag: "mvoq5wehljnqqs"
via: 1.1 varnish, 1.1 varnish
accept-ranges: bytes
date: Sun, 30 Mar 2025 21:59:12 GMT
age: 236
x-served-by: cache-ewr-kewr1740069-EWR, cache-lon420094-LON
x-cache: HIT, HIT
x-cache-hits: 2, 1
x-timer: S1743371952.295567,VS0,VE2
vary: Accept-Encoding, Accept-Encoding
strict-transport-security: max-age=31557600
alt-svc: h3=":443";ma=86400,h3-29=":443";ma=86400,h3-27=":443";ma=86400
content-length: 1236032

elbaih@abolebda:~$ 

```
